### PR TITLE
Cura Definition for Flasforge DreamerNX

### PR DESCRIPTION
**Pull request by KeltE**
Cura Definition for Flasforge DreamerNX

### CODE REVIEW

File: resources/definitions/flashforge_dreamer_nx.def.json [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        "visible": true,
+        "author": "Egon",
+        "manufacturer": "Flashforge",
+        "file_formats": "text/x-gcode",
+        "platform": "FlashForge_DreamerNX.obj",
+        "platform_offset": [0, 0, 0],
+        "quality_definition": "flashforge_base",
+        "machine_extruder_trains":
+        {
+            "0": "ff_dreamernx_extruder_0"
```
Comment (konskarm) [inline resources/definitions/flashforge_dreamer_nx.def.json:15]: Minor thing, but please rename the extruder to include the full name of the printer
```suggestion
            "0": "flashforge_dreamernx_extruder_0"
```

Also rename the file `resources/extruders/ff_dreamernx_extruder_0.def.json` to `resources/extruders/flashforge_dreamernx_extruder_0.def.json`

File: resources/definitions/flashforge_base.def.json [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        "has_materials": true,
+        "has_variants": true,
+        "has_machine_quality": true,
+        "preferred_variant_name": "0.4mm Nozzle",
+        "preferred_quality_type": "normal",
+        "preferred_material": "generic_pla",
+        "variants_name": "Nozzle Size"
+    },
+    "overrides": {
+		"machine_heated_bed": {"default_value": true},
```
Comment (konskarm) [inline resources/definitions/flashforge_base.def.json:24]:
```suggestion
        "machine_heated_bed": {"default_value": true},
```

File: resources/definitions/flashforge_dreamer_nx.def.json [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        }
+    },
+
+    "overrides": {
+        "machine_name": { "default_value": "Dreamer NX" },
+        "machine_width": {"default_value": 230},
+        "machine_height": {"default_value": 140},
+        "machine_depth": {"default_value": 150},
+        "machine_center_is_zero": {"default_value": true},
+		"machine_heated_bed": {"default_value": true},
```
Comment (konskarm) [inline resources/definitions/flashforge_dreamer_nx.def.json:25]: Mix of tabs and spaces
```suggestion
        "machine_heated_bed": {"default_value": true},
```

File: resources/definitions/flashforge_dreamer_nx.def.json [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        "machine_max_feedrate_z": { "value": 10 },
+        "machine_max_feedrate_e": { "value": 50 },
+        "machine_max_acceleration_x": { "value": 1500 },
+        "machine_max_acceleration_y": { "value": 1500 },
+        "machine_max_acceleration_z": { "value": 500 },
+        "machine_max_acceleration_e": { "value": 5000 },
+        "machine_acceleration": { "value": 500 },
+        "machine_max_jerk_xy": { "value": 10 },
+        "machine_max_jerk_z": { "value": 0.4 },
+        "machine_max_jerk_e": { "value": 5 },
```
Comment (konskarm) [inline resources/definitions/flashforge_dreamer_nx.def.json:37]: Reduntant values. They are already inherited by the base machine profile.
```suggestion
```

File: resources/definitions/flashforge_base.def.json [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    "inherits": "fdmprinter",
+    "metadata": {
+        "visible": false,
+        "author": "Egon",
+        "manufacturer": "Flashforge",
+        "file_formats": "text/x-gcode",
+        "first_start_actions": ["MachineSettingsAction"],
+        "machine_extruder_trains":
+        {
+            "0": "ff_dreamernx_extruder_0"
```
Comment (konskarm) [inline resources/definitions/flashforge_base.def.json:13]: Minor thing, but please rename the extruder to include the full name of the printer
```suggestion
            "0": "flashforge_dreamernx_extruder_0"
```

Also rename the file `resources/extruders/ff_dreamernx_extruder_0.def.json` to `resources/extruders/flashforge_dreamernx_extruder_0.def.json`

File: resources/definitions/flashforge_dreamer_nx.def.json [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        "author": "Egon",
+        "manufacturer": "Flashforge",
+        "file_formats": "text/x-gcode",
+        "platform": "FlashForge_DreamerNX.obj",
+        "platform_offset": [0, 0, 0],
+        "quality_definition": "flashforge_base",
+        "machine_extruder_trains":
+        {
+            "0": "ff_dreamernx_extruder_0"
+        }
```
Comment (konskarm) [inline resources/definitions/flashforge_dreamer_nx.def.json:16]: Reduntant, as it is already inherited by the `flashforge_base.def.json`.
```suggestion
```

**Review by konskarm — CHANGES_REQUESTED** [review]

File: resources/definitions/flashforge_dreamer_nx.def.json (hunk 6 above)
Comment (KeltE) [reply resources/definitions/flashforge_dreamer_nx.def.json:16]: This is my first time. Thank you for pointing out the errors. I'll fix the mistakes tomorrow.
Thanks again!

**Comment by konskarm** [comment]
I may have misled you a bit. Along with commit https://github.com/Ultimaker/Cura/pull/9303/commits/d2b55f614cf5467772f1100c01f48930096163ee you also needed to change the actual name of the file `resources/extruders/ff_dreamernx_extruder_0.def.json` into `resources/extruders/flashforge_dreamernx_extruder_0.def.json`. Now the tests cannot find that file.

**Review by konskarm — CHANGES_REQUESTED** [review]
I may have misled you a bit. Along with commit https://github.com/Ultimaker/Cura/pull/9303/commits/d2b55f614cf5467772f1100c01f48930096163ee you also needed to change the actual name of the file `resources/extruders/ff_dreamernx_extruder_0.def.json` into `resources/extruders/flashforge_dreamernx_extruder_0.def.json`. Now the tests cannot find that file.

Sorry for that, but unfortunately GitHub doesn't allow me to suggest filename changes (AFAIK).

**Comment by KeltE** [comment]
I already understood this when I changed your recommendations. I already changed the file name. Hopefully he didn't make any more mistakes. Still confusing Github for me. The terms of all is not yet well understood.

**Comment by konskarm** [comment]
Hmm then this change must have not been commited, because the filename is still [`resources/extruders/ff_dreamernx_extruder_0.def.json`](https://github.com/Ultimaker/Cura/pull/9303/files#diff-0ca013d63e37f647a9adcc6cc62ff4e7d2fcafabd4dda72183bbe1dfa8fc0cf2).

**Comment by KeltE** [comment]
Maybe I made a mistake.
 I only changed the file name in the printer definition folder. Are changes to my Cura branch not automatically updated?

File: resources/definitions/flashforge_dreamer_nx.def.json [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    "name": "Dreamer NX",
+    "inherits": "flashforge_base",
+    "metadata": {
+        "visible": true,
+        "author": "Egon",
+        "manufacturer": "Flashforge",
+        "file_formats": "text/x-gcode",
+        "platform": "FlashForge_DreamerNX.obj",
+        "platform_offset": [0, 0, 0],
+        "quality_definition": "flashforge_base",
```
Comment (konskarm) [inline resources/definitions/flashforge_dreamer_nx.def.json:12]: I think this comma is the cause of the failing tests.
```suggestion
        "quality_definition": "flashforge_base"
```

**Comment by KeltE** [comment]
Thank you very much!  Now the check is successful.

**Comment by konskarm** [comment]
Excellent. I also tested it and it slices without issues so I am merging it. The machine will be available to be added in Cura 4.9.

**Comment by KeltE** [comment]
Great! Thank you very much for your help.

**Comment by HadiSDev** [comment]
@KeltE Willl this one work for the Dreamer as well?

**Comment by KeltE** [comment]
> @KeltE Willl this one work for the Dreamer as well?

Unfortunately, there is no experience with a twin-extruder printer, so I can't say exactly. DreamerNX is a single extruder printer. The dreamer then needs to make his own definition file and both extruders his own extruder file.

**Pull request merged**